sdk/python/tests/test_materialize.py

```diff
@@ -0,0 +1,168 @@
+            input=data_source,
+        )
+
+    def run_materialization_test(self, fs: FeatureStore, fv: FeatureView):
+        now = datetime.utcnow()
+        # Run materialize()
+        # use both tz-naive & tz-aware timestamps to test that they're both correctly handled
+        start_date = (now - timedelta(hours=5)).replace(tzinfo=utc)
+        end_date = now - timedelta(hours=2)
+        fs.materialize([fv.name], start_date, end_date)
+
+        # check result of materialize()
+        response_dict = fs.get_online_features(
+            [f"{fv.name}:value"], [{"driver_id": 1}]
+        ).to_dict()
+        assert abs(response_dict[f"{fv.name}__value"][0] - 0.3) < 1e-6
+
+        # check prior value for materialize_incremental()
+        response_dict = fs.get_online_features(
+            [f"{fv.name}:value"], [{"driver_id": 3}]
+        ).to_dict()
+        assert abs(response_dict[f"{fv.name}__value"][0] - 4) < 1e-6
+
+        # run materialize_incremental()
+        fs.materialize_incremental(
+            [fv.name], now - timedelta(seconds=0),
+        )
+
+        # check result of materialize_incremental()
+        response_dict = fs.get_online_features(
+            [f"{fv.name}:value"], [{"driver_id": 3}]
+        ).to_dict()
+        assert abs(response_dict[f"{fv.name}__value"][0] - 5) < 1e-6
+
+    @pytest.mark.integration
+    @pytest.mark.parametrize(
+        "bigquery_source", [lazy_fixture("bigquery_source_fixture")],
+    )
+    def test_bigquery_to_datastore_correctness(self, bigquery_source: BigQuerySource):
+        # create FeatureView
+        fv = self.get_feature_view(bigquery_source)
+        with tempfile.TemporaryDirectory() as repo_dir_name:
+            config = RepoConfig(
+                registry=str(Path(repo_dir_name) / "registry.db"),
+                project=f"test_bq_correctness_{uuid.uuid4()}",
+                provider="gcp",
+            )
+            fs = FeatureStore(config=config)
+            fs.apply([fv])
+
+            self.run_materialization_test(fs, fv)
+
+    @pytest.mark.parametrize(
+        "file_source", [lazy_fixture("file_source_fixture")],
+    )
+    def test_local_correctness(self, file_source: FileSource):
+        # create FeatureView
+        fv = self.get_feature_view(file_source)
+        with tempfile.TemporaryDirectory() as repo_dir_name, tempfile.TemporaryDirectory() as data_dir_name:
+            config = RepoConfig(
+                registry=str(Path(repo_dir_name) / "registry.db"),
+                project=f"test_bq_correctness_{str(uuid.uuid4()).replace('-', '')}",
+                provider="local",
+                online_store=OnlineStoreConfig(
+                    local=LocalOnlineStoreConfig(
+                        path=str(Path(data_dir_name) / "online_store.db")
+                    )
+                ),
+            )
+            fs = FeatureStore(config=config)
+            fs.apply([fv])
+
+            self.run_materialization_test(fs, fv)
```

Comment:
This is a bit of a nitpick, but what do you think of restructuring the tests a little bit? I dont really see the need for a class and methods here when we can use functions. I'm worried that the class will create unnecessary scoping or tests will evolve to influence each other over time. I also see you are mixing graph building in a push (`self.run_materialization_test(fs, fv)`) and pull (`@pytest.mark.parametrize`) manner. We should probably just pick one of the two, and the pull based graph building seems cleaner to me.

something like

def other_functions()

def prep_bq_fs()
  do all bq setup

def prep_local_fs()
  do all local setup

@pytest.mark.parametrize( depends on prep_bq_fs and prep_local_fs)
def test_materialization()

Reply:
Most of this makes sense, but the problem with
> @pytest.mark.parametrize( depends on prep_bq_fs and prep_local_fs)
def test_materialization()

is that I either have to make the entire test function as integration test or unit test. There's no way to selectively mark prep_bq_fs as integration and prep_local_fs as unit test (I just tried this locally).

Reply:
@woop I went ahead with the push based approach. Basically make everything functions that we call in order to not duplicate anything